code/eyetracking/eyetrackingrun.py

```diff
@@ -0,0 +1,961 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+import re
+import os
+import json
+from pathlib import Path
+import string
+from collections import defaultdict
+from typing import Optional, List, Tuple, Union
+
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+from pyedfread import edf, edfread
+
+
+# EyeLink calibration coordinates from https://www.sr-research.com/calibration-coordinate-calculator/
+EYELINK_CALIBRATION_COORDINATES = [
+    (400, 300), (400, 51), (400, 549), (48, 300), (752, 300), (48, 51), (752, 51), (48, 549), (752, 549), (224, 176), (576, 176), (224, 424), (576, 424)
+]
+
+EYE_CODE_MAP = defaultdict(
+    lambda: "unknown", {"R": "right", "L": "left", "RL": "both"}
+)
+RIGHT_EYE_COLUMNS = {
+    "time": "eye_timestamp",
+    "gx_right": "eye1_x_coordinate",
+    "gy_right": "eye1_y_coordinate",
+    "pa_right": "eye1_pupil_size",
+    "px_right": "eye1_pupil_x_coordinate",
+    "py_right": "eye1_pupil_y_coordinate",
+    "hx_right": "eye1_head_x_coordinate",
+    "hy_right": "eye1_head_y_coordinate",
+    "rx": "screen_pixel_per_degree_x",
+    "ry": "screen_pixel_per_degree_y",
+    "gxvel_right": "eye1_x_velocity",
+    "gyvel_right": "eye1_x_velocity",
+    "hxvel_right": "eye1_head_x_velocity",
+    "hyvel_right": "eye1_head_y_velocity",
+    "rxvel_right": "eye1_raw_x_velocity",
+    "ryvel_right": "eye1_raw_y_velocity",
+    "fgxvel": "eye1_fast_gaze_x_velocity",
+    "fgyvel": "eye1_fast_gaze_y_velocity",
+    "fhxyvel": "eye1_fast_head_x_velocity",
+    "fhyvel": "eye1_fast_head_y_velocity",
+    "frxyvel": "eye1_fast_raw_x_velocity",
+    "fryvel": "eye1_fast_raw_y_velocity",
+}
+
+
+class EyeTrackingRun:
+    """
+    Class representing an instance of eye tracking data.
+
+    Parameters
+    ----------
+    session : int
+        Session ID.
+    task_name : str
+        Task name.
+    participant : int
+        Participant ID.
+    samples : pd.DataFrame
+        DataFrame containing eye tracking samples.
+    events : pd.DataFrame
+        DataFrame containing eye tracking events.
+    messages : pd.DataFrame
+        DataFrame containing eye tracking messages.
+    message_first_trigger : str
+        Message that signals the first trigger of the run.
+    screen_resolution : Tuple[int, int]
+        Screen resolution as a tuple of two integers.
+
+    pe : str, optional
+        Phase encoding direction (default is an empty string).
+
+    Examples
+    --------
+    This class can be instantiated as follows:
+
+    ```python
+    et_run = EyeTrackingRun(
+        session=1,
+        task_name="rest",
+        participant=001,
```

Comment:
It looks like session, task_name, participant, and pe are extremely customized to our use-case. Why not passing the path to the corresponding BIDS file this ET recording corresponds to?

Reply:
BTW, participant is missing the quotes:
```suggestion
        session=1,
        task_name="rest",
        participant="001",
```

Reply:
(Edited)
The reason behind passing session, task_name, participant, and pe as arguments is that not only the path but also the filename contains those variables. Having to give the entire filename as an argument seemed risky to me, especially with the '_' and '-' that can easily lead to typos. (Cf #235  and the comment [here](https://github.com/TheAxonLab/hcph-sops/pull/340#discussion_r1385202505). But I agree it's not ideal, so let me know your thoughts. If you believe path + name is still a better approach, I'll go for that. Maybe this could go in a new PR?